lib/proxy_ssh/proxy.go

```diff
@@ -0,0 +1,282 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package proxy_ssh
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"sync"
+
+	"golang.org/x/crypto/ssh"
+
+	"github.com/adobe/aquarium-fish/lib/fish"
+	"github.com/adobe/aquarium-fish/lib/log"
+	"github.com/adobe/aquarium-fish/lib/openapi/types"
+)
+
+func init() {
+	log.Info("The Fish SSH proxy is a re-implementation of Remco Verhoef's via MIT licensed https://github.com/dutchcoders/sshproxy")
+}
+
+type ProxyAccess struct {
+	fish         *fish.Fish
+	serverConfig *ssh.ServerConfig
+	// Keys: remote address from ssh connection
+	// Values: tuple (username, password)
+	// [all strings]
+	sessions sync.Map
+}
+
+// Stored in ProxyAccess::sessions.
+type SessionRecord struct {
+	ResourceAccessor *types.ResourceAccess
+	RemoteAddr       net.Addr
+	// Username   string
+	// Password   string
+}
+
+func (p *ProxyAccess) serveConnection(conn net.Conn, serverConfig *ssh.ServerConfig) error {
+	// Establish the initial connection.
+	serverConn, serverChannels, serverReqs, err := ssh.NewServerConn(conn, serverConfig)
+	if err != nil {
+		return log.Errorf("Failed to establish new server connection with %s: %v", conn.RemoteAddr(), err)
+	}
+	defer serverConn.Close()
+	go ssh.DiscardRequests(serverReqs)
+
+	// Pop off the session information from our passwordCallback map immediately
+	// as any new connections will pass through that callback and add it.
+	remoteAddrString := conn.RemoteAddr().String()
+	value, loaded := p.sessions.LoadAndDelete(remoteAddrString)
+	if value == nil || !loaded {
+		return log.Errorf("Unable to load session record for %q", remoteAddrString)
+	}
+
+	// Go find the resource via its UID.
+	session_record, ok := value.(SessionRecord)
+	if !ok {
+		return log.Errorf("Critical error retrieving session record (invalid type conversion).")
+	}
+	resource, err := p.fish.ResourceGet(session_record.ResourceAccessor.ResourceUID)
+	if err != nil {
+		return log.Errorf("Unable to retrieve resource: %v", err)
+	}
+
+	// If the resource was not created with the authentication username and
+	// password then they will both be the empty string, meaning we cannot
+	// actually login to anything.
+	if resource.Authentication.Username == "" && resource.Authentication.Password == "" {
+		// TODO: in the future, this may not be an appropriate error.
+		return log.Errorf("Resource Authentication not provided")
+	}
+
+	// Create the destination end of the proxy to the remote host.
+	remoteAddr := fmt.Sprintf("%s:22", resource.IpAddr)
+	remoteConfig := &ssh.ClientConfig{
+		User: resource.Authentication.Username,
+		Auth: []ssh.AuthMethod{
+			ssh.Password(resource.Authentication.Password),
+		},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
```

Comment:
It's not clear to me that this is something we can actually fix.  See details: https://github.com/adobe/aquarium-fish/security/code-scanning/8

we don't know what the approved host key actually is do we?

Reply:
We need to ignore that at least for now - there is no easy way to properly ensure the ssh connection by the key because each instance generates the new one (during the image build cleaning step we remove the genrated host key from the system so it creates the new one on each run). Later we could have a logic, that will try to connect to the instance via ssh on allocation completed - but that future solution have a lot of "if"'s.

Reply:
I'm going to see if I can find a workaround, but this may become a showstopper / require the removal of CodeQL.

- No ability to suppress anything: https://github.com/github/codeql/issues/11427
- There is an [additional suppression action](https://github.com/advanced-security/dismiss-alerts?tab=readme-ov-file#features-and-limitations), but it comes with a pretty big cost...

[This CVE](https://codeql.github.com/codeql-query-help/go/go-insecure-hostkeycallback/) is what we're trying to workaround, seems like a dummy function is also going to get flagged.  Going to try a few different versions.

Reply:
No need to spend time on that - I can dismiss the codeql issues, so I did.